### PR TITLE
adding batch_input to transit.sign_data #988

### LINKED ISSUE #988: batch_input attribute is not available for sign_data
it seems that **Transit.sign_data** doesn't accept a batch_input attribute like in the **Transit.encrypt_data** even though the Vault API accepts batch_input for **/transit/sign/:name(/:hash_algorithm)**

### PR DESCRIPTION
**Pull request by deidax**
Resolves #988 by adding the `batch_input` to the `transit sign_data method` and updating the params.

~Resolve the issue #989 by checking if the value of hash_algorithm is None `hash_algorithm = '' if not hash_algorithm else hash_algorithm` and updating the api_path `"/v1/{mount_point}/sign/{name}/{hash_algorithm}"`~

### CODE REVIEW

File: Pipfile
Comment (briantist) [inline Pipfile:1]: This file doesn't appear related to this change, could you explain what it is or remove it?

File: hvac/api/secrets_engines/transit.py
Comment (briantist) [inline hvac/api/secrets_engines/transit.py:760]: Please move this to the end of the argument list, otherwise it will break folks using positional arguments.

File: hvac/api/secrets_engines/transit.py
Comment (briantist) [inline hvac/api/secrets_engines/transit.py:877]:
```suggestion
                    "batch_input": batch_input,
```

File: hvac/api/secrets_engines/transit.py
Comment (briantist) [inline hvac/api/secrets_engines/transit.py:888]:
```suggestion
            hash_algorithm=hash_algorithm,
```

**Review by briantist — COMMENTED** [review]
thanks for submitting this! I've left a few small comments, but I still need to look at the underlying issues more deeply. I'm a little overloaded at the moment, but I will get back to it at some point.

File: Pipfile
Comment (deidax) [reply Pipfile:1]: thank you for pointing this out, the Pipfile is not related to this change

**Comment by briantist** [comment]
Hello @deidax ! 

The `hvac` project is simplifying its developer workflow and removing the `develop` branch. Please be aware for future PRs that the project's default branch is now `main`.

Since your PR was opened before this change, I have: 
- updated your PR to merge into the project's `main` branch
- rebased your branch against `main` and force-pushed it (**please pull before making any further changes**)

I will come back to review the actual content as soon as I can, thanks for your patience!

File: hvac/api/secrets_engines/transit.py
Comment (briantist) [inline hvac/api/secrets_engines/transit.py:766]: This needs a docstring entry below so that it will be documented properly. The documentation for the API should help with the description: https://developer.hashicorp.com/vault/api-docs/secret/transit#batch_input-4

File: hvac/api/secrets_engines/transit.py
Comment (briantist) [inline hvac/api/secrets_engines/transit.py:885]: The description for this parameter says that it can _also_ be specified in the URL, not that it must be. The method seems to already submit it in parameters, so I'm not certain why we would require it to be in the URL. Let's continue this in the issue report: 
- https://github.com/hvac/hvac/issues/989

If you can update the tests in a way that demonstrates the issue, it would also be very helpful.

**Review by briantist — COMMENTED** [review]
Thanks for submitting this! As commented below, I'd like to get more information on the `hash_algorithm` issue.

For `batch_input`, it needs docs, and I'd really like to see the tests updated for this, is that something you could take a look at?

File: hvac/api/secrets_engines/transit.py
Comment (deidax) [reply hvac/api/secrets_engines/transit.py:885]: > The description for this parameter says that it can _also_ be specified in the URL, not that it must be. The method seems to already submit it in parameters, so I'm not certain why we would require it to be in the URL. Let's continue this in the issue report:
> 
> * [transit sign_data doesn't take the hash_algorithm in consideration #989](https://github.com/hvac/hvac/issues/989)
> 
> If you can update the tests in a way that demonstrates the issue, it would also be very helpful.

hi @briantist , that's true but I got this issue when I was working on a project that needed the hash algorithm to be passed in the URL and after looking at Vault documentation here https://developer.hashicorp.com/vault/api-docs/secret/transit#hash-data I noticed that the hash algorithm is passed in the encryption URL. The changes I made here worked for my case.

**Comment by briantist** [comment]
@deidax could you please revert the merge of the `develop` branch into the PR? 
Both of these commits should be reverted or removed via interactive rebase:
- https://github.com/hvac/hvac/pull/990/commits/38378d695cb7fcd3a887929024fba1b8e68b03c8
- https://github.com/hvac/hvac/pull/990/commits/eca6563cbab62c3c77207d42027c4a285706d217

As well as the 4 commits shown here that have only me as the author:
- https://github.com/hvac/hvac/pull/990/commits

See here for why: https://github.com/hvac/hvac/pull/990#issuecomment-1595858305

File: hvac/api/secrets_engines/transit.py
Comment (briantist) [reply hvac/api/secrets_engines/transit.py:885]: Let's continue this in the issue report. If you prefer, you can also change this PR to only add `batch_input` and then a separate PR could be used for the other thing once we figure out what's going on, or we can work that out first, whatever works for you.

**Comment by deidax** [comment]
> @deidax could you please revert the merge of the `develop` branch into the PR? Both of these commits should be reverted or removed via interactive rebase:
> 
> * [38378d6](https://github.com/hvac/hvac/commit/38378d695cb7fcd3a887929024fba1b8e68b03c8)
> * [eca6563](https://github.com/hvac/hvac/commit/eca6563cbab62c3c77207d42027c4a285706d217)
> 
> As well as the 4 commits shown here that have only me as the author:
> 
> * https://github.com/hvac/hvac/pull/990/commits
> 
> See here for why: [#990 (comment)](https://github.com/hvac/hvac/pull/990#issuecomment-1595858305)
> 
> If you need help, I can try to do this and force push your branch.

@briantist Yes I would appreciate it if you can do it, because I'm a little lost here 😅

File: hvac/api/secrets_engines/transit.py
Comment (deidax) [reply hvac/api/secrets_engines/transit.py:885]: @briantist let's change this PR to `batch_input` for now.

**Comment by deidax** [comment]
> Thanks for submitting this! As commented below, I'd like to get more information on the `hash_algorithm` issue.
> 
> For `batch_input`, it needs docs, and I'd really like to see the tests updated for this, is that something you could take a look at?

@briantist Before running any tests, please consider the following: According to the vault documentation for the `transit.sign_data` function, if `batch_input` is set, `hash_input` should be ignored. Currently, the function does not handle this case properly, as `hash_input` is a required parameter. To use `batch_input` instead, the user should set `hash_input=None`.

File: hvac/api/secrets_engines/transit.py
Comment (deidax) [inline hvac/api/secrets_engines/transit.py:872]: To ensure compatibility, I included the following line to set the value of `hash_input` to `None` when `batch_input` is set, as both values cannot be simultaneously used.

**Comment by briantist** [comment]
> > @deidax could you please revert the merge of the `develop` branch into the PR? Both of these commits should be reverted or removed via interactive rebase:
> > 
> > * [38378d6](https://github.com/hvac/hvac/commit/38378d695cb7fcd3a887929024fba1b8e68b03c8)
> > * [eca6563](https://github.com/hvac/hvac/commit/eca6563cbab62c3c77207d42027c4a285706d217)
> > 
> > As well as the 4 commits shown here that have only me as the author:
> > 
> > * https://github.com/hvac/hvac/pull/990/commits
> > 
> > See here for why: [#990 (comment)](https://github.com/hvac/hvac/pull/990#issuecomment-1595858305)
> > If you need help, I can try to do this and force push your branch.
> 
> Yes I would appreciate it if you can do it

Sure, I'm going to attempt this now. Since this operation can lose data, you may want to create a new branch off of your current one locally in case you need to recover commits later.

**Comment by briantist** [comment]
> > > @deidax could you please revert the merge of the `develop` branch into the PR? Both of these commits should be reverted or removed via interactive rebase:
> > > 
> > > * [38378d6](https://github.com/hvac/hvac/commit/38378d695cb7fcd3a887929024fba1b8e68b03c8)
> > > * [eca6563](https://github.com/hvac/hvac/commit/eca6563cbab62c3c77207d42027c4a285706d217)
> > > 
> > > As well as the 4 commits shown here that have only me as the author:
> > > 
> > > * https://github.com/hvac/hvac/pull/990/commits
> > > 
> > > See here for why: [#990 (comment)](https://github.com/hvac/hvac/pull/990#issuecomment-1595858305)
> > > If you need help, I can try to do this and force push your branch.
> > 
> > 
> > Yes I would appreciate it if you can do it
> 
> Sure, I'm going to attempt this now. Since this operation can lose data, you may want to create a new branch off of your current one locally in case you need to recover commits later.

@deidax I believe this was successful! I also pushed an additional commit where I ran black to fix the formatting/lint error.

**For this PR, please do not use the GitHub UI update branch button. You can ignore the message telling you that your branch is behind.**

What's happening is that you have a `develop` branch which tracks our `develop` branch, but we're not using that one anymore. I previously _rebased_ your `develop` branch to actually be based on our `main` branch.

I think this will not be a problem for your future PRs (I hope there will be more 😁), which should be based off the `main` branch to begin with.

I'm very sorry for the confusion about this, I know it's not a good experience especially for a first-time contributor to a project, it was just bad timing on my part.

File: hvac/api/secrets_engines/transit.py
Comment (briantist) [reply hvac/api/secrets_engines/transit.py:895]: Since we are changing this PR to be for `batch_input` only, please undo these `hash_algorithm` changes for now

File: hvac/api/secrets_engines/transit.py
Comment (briantist) [reply hvac/api/secrets_engines/transit.py:872]: Thanks for pointing this out. I am thinking about whether we should continue to have `hash_input` be a required parameter (argument) in the function, or we should default it to `None`, and then have code to raise an exception if `hash_input is None and batch_input is None`.

Basically, I am thinking it might be an awkward user experience when using `batch_input` if you are forced to explicitly set `hash_input=None`.

Since you would presumably be using this exact scenario, what do you think about that?

Comment (deidax) [reply hvac/api/secrets_engines/transit.py:872]: I think it would be better set both parameters to None by default. In case the user fails to provide a value for any of the parameters, an exception can be raised with message guiding the user to set the necessary values.

File: hvac/api/secrets_engines/transit.py
Comment (briantist) [inline hvac/api/secrets_engines/transit.py:814]:
```suggestion
            This parameter is mutually exclusive with the ``hash_input`` parameter, but one of them must be supplied. 
            If both are set, or neither are set, an exception will be raised.
            Responses are returned in the ``batch_results`` array component of the ``data`` element of the response.
            Any batch output will preserve the order of the batch input.
            If the input data value of an item is invalid, the corresponding item in the ``batch_results``
            will have the key ``error`` with a value describing the error.
```

Updating to match new behavior. Please add something similar to the first two lines to the description for `hash_input` as wel.

**Review by briantist — COMMENTED** [review]
Just some wording and formatting suggestions

File: hvac/api/secrets_engines/transit.py
Comment (briantist) [inline hvac/api/secrets_engines/transit.py:779]: reordering a little to put the actual description first
```suggestion
        :param hash_input: Specifies the base64 encoded input data.
            This parameter is mutually exclusive with the ``batch_results`` parameter, but one of them must be supplied. 
            If both are set, or neither are set, an exception will be raised.
```

File: hvac/api/secrets_engines/transit.py
Comment (briantist) [inline hvac/api/secrets_engines/transit.py:817]: some more re-ordering to put description first and align with the other param.
```suggestion
        :param batch_input: Specifies a list of items for processing.
            Any batch output will preserve the order of the batch input.
            If the input data value of an item is invalid, the corresponding item in the ``batch_results``
            will have the key ``error`` with a value describing the error.
            This parameter is mutually exclusive with the ``hash_input`` parameter, but one of them must be supplied. 
            If both are set, or neither are set, an exception will be raised.
            Responses are returned in the ``batch_results`` array component of the ``data`` element of the response.
```

**Review by briantist — COMMENTED** [review]
a few more docs re-ordering, I'll go ahead and commit these, and if you want to change anything with the edits, feel free to push more commits

**Review by briantist — APPROVED** [review]
This looks good to me! Thanks so much for submitting and sticking through the edits and suggestions.

Since I was the last committer, I'll wait for you to look at those changes before merging. Let me know if it all looks good to you.

**Comment by deidax** [comment]
> This looks good to me! Thanks so much for submitting and sticking through the edits and suggestions.
> 
> Since I was the last committer, I'll wait for you to look at those changes before merging. Let me know if it all looks good to you.

@briantist yeah it looks good to me. thank you

**Pull request merged**
